File: FBAudio-iOS/FBAudio/Views/Legacy/SamplePlayer.swift
import Foundation
import AVFoundation
import Combine

/// A/B player for the Digital Legacy sample: the original and the remastered
/// recording of one chapter are both prepared up front and play in lockstep,
/// with the inactive one muted — so switching version is instant, no
/// re-buffering. A demo, not background listening: paused when the page is
/// left, released with the view. Mirrors Android's SamplePlayer.
@MainActor
final class SamplePlayer: ObservableObject {
    @Published private(set) var useRemaster = true
    @Published private(set) var isPlaying = false
    @Published private(set) var isBuffering = false
    @Published private(set) var positionMs: Int64 = 0
    @Published private(set) var durationMs: Int64 = 0
    @Published private(set) var isReady = false

    private var original = AVPlayer()
    private var remaster = AVPlayer()
    private var timer: Timer?
    private var ticks = 0

    private var active: AVPlayer { useRemaster ? remaster : original }
    private var shadow: AVPlayer { useRemaster ? original : remaster }

    /// Prepare both versions (buffering starts immediately) without playing.
    func load(originalUrl: String, remasterUrl: String, startWithRemaster: Bool, knownDurationMs: Int64) {
        guard let o = URL(string: originalUrl), let r = URL(string: remasterUrl) else { return }
        useRemaster = startWithRemaster
        durationMs = knownDurationMs
        original.replaceCurrentItem(with: AVPlayerItem(url: o))
        remaster.replaceCurrentItem(with: AVPlayerItem(url: r))
        original.automaticallyWaitsToMinimizeStalling = false
        remaster.automaticallyWaitsToMinimizeStalling = false
        // Start filling both buffers now so Play is immediate
        original.preroll(atRate: 1.0)
        remaster.preroll(atRate: 1.0)
        applyVolumes()
        isReady = true
        startTimer()
    }

    func play() {
        try? AVAudioSession.sharedInstance().setActive(true)
        if let d = active.currentItem?.duration.seconds, d.isFinite, active.currentTime().seconds >= d - 0.5 {
            original.seek(to: .zero); remaster.seek(to: .zero)
        }
        shadow.seek(to: active.currentTime(), toleranceBefore: .zero, toleranceAfter: .zero)
        original.play()
        remaster.play()
        tick()
    }

    func pause() {
        original.pause()
        remaster.pause()
        tick()
    }

    func togglePlayPause() { isPlaying ? pause() : play() }

    /// Swap what's audible; the newly audible player is aligned to the current position.
    func setVersion(_ remastered: Bool) {
        guard remastered != useRemaster else { return }
        let position = active.currentTime()
        useRemaster = remastered
        active.seek(to: position, toleranceBefore: .zero, toleranceAfter: .zero)
        applyVolumes()
    }

    func seek(toFraction fraction: Double) {
        guard durationMs > 0 else { return }
        let seconds = Double(durationMs) / 1000 * min(max(fraction, 0), 1)
        let time = CMTime(seconds: seconds, preferredTimescale: 600)
        original.seek(to: time, toleranceBefore: .zero, toleranceAfter: .zero)
        remaster.seek(to: time, toleranceBefore: .zero, toleranceAfter: .zero)
        tick()
    }

    private func applyVolumes() {
        original.volume = useRemaster ? 0 : 1
        remaster.volume = useRemaster ? 1 : 0
    }

    private func startTimer() {
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.tick() }
        }
    }

    private func tick() {
        isPlaying = active.timeControlStatus != .paused
        isBuffering = active.timeControlStatus == .waitingToPlayAtSpecifiedRate
        let pos = active.currentTime().seconds
        positionMs = pos.isFinite ? Int64(pos * 1000) : 0
        if let d = active.currentItem?.duration.seconds, d.isFinite, d > 0 { durationMs = Int64(d * 1000) }
        // Pull the muted player back into step if it has drifted
        ticks += 1
        if isPlaying, ticks % 10 == 0 {
            let drift = abs(shadow.currentTime().seconds - active.currentTime().seconds)
            if drift.isFinite, drift > 0.75 {
                shadow.seek(to: active.currentTime(), toleranceBefore: .zero, toleranceAfter: .zero)
            }
        }
    }

    deinit {
        timer?.invalidate()
        original.pause()
        remaster.pause()
    }
}

File: FBAudio-Android/app/src/main/java/com/fba/app/ui/legacy/SamplePlayer.kt
package com.fba.app.ui.legacy

import android.content.Context
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer

/**
 * A/B player for the Digital Legacy sample: the original and the remastered
 * recording of one chapter are both prepared up front and play in lockstep,
 * with the inactive one muted — so switching version is instant, no
 * re-buffering. Lives in the UI process (it's a demo, not background listening)
 * and is released when the screen goes away.
 */
class SamplePlayer(context: Context) {
    private val original: ExoPlayer = ExoPlayer.Builder(context).build().apply {
        // Only one of the pair asks for audio focus — a second request would pause the first.
        setAudioAttributes(AudioAttributes.Builder().setUsage(C.USAGE_MEDIA).setContentType(C.AUDIO_CONTENT_TYPE_SPEECH).build(), true)
        setHandleAudioBecomingNoisy(true)
    }
    private val remaster: ExoPlayer = ExoPlayer.Builder(context).build().apply {
        setAudioAttributes(AudioAttributes.Builder().setUsage(C.USAGE_MEDIA).setContentType(C.AUDIO_CONTENT_TYPE_SPEECH).build(), false)
    }

    var useRemaster: Boolean = true
        private set

    private val active: ExoPlayer get() = if (useRemaster) remaster else original
    private val shadow: ExoPlayer get() = if (useRemaster) original else remaster

    val isPlaying: Boolean get() = active.playWhenReady && active.playbackState != Player.STATE_ENDED
    val isBuffering: Boolean get() = active.playbackState == Player.STATE_BUFFERING && active.playWhenReady
    val positionMs: Long get() = active.currentPosition.coerceAtLeast(0)
    val durationMs: Long get() = active.duration.takeIf { it != C.TIME_UNSET }?.coerceAtLeast(0) ?: 0
    val isEnded: Boolean get() = active.playbackState == Player.STATE_ENDED

    /** Prepare both versions (buffering starts immediately) without playing. */
    fun load(originalUrl: String, remasterUrl: String, startWithRemaster: Boolean) {
        useRemaster = startWithRemaster
        original.setMediaItem(MediaItem.fromUri(originalUrl))
        remaster.setMediaItem(MediaItem.fromUri(remasterUrl))
        original.prepare()
        remaster.prepare()
        applyVolumes()
    }

    fun play() {
        if (isEnded) { original.seekTo(0); remaster.seekTo(0) }
        // Start the shadow at the active position so the two stay in step
        shadow.seekTo(active.currentPosition)
        original.play()
        remaster.play()
    }

    fun pause() {
        original.pause()
        remaster.pause()
    }

    fun togglePlayPause() = if (isPlaying) pause() else play()

    /** Swap what's audible; the newly audible player is aligned to the current position. */
    fun setVersion(remastered: Boolean) {
        if (remastered == useRemaster) return
        val position = active.currentPosition
        useRemaster = remastered
        active.seekTo(position)
        applyVolumes()
    }

    fun seekTo(positionMs: Long) {
        original.seekTo(positionMs)
        remaster.seekTo(positionMs)
    }

    /** Called periodically: pull the muted player back into step if it has drifted. */
    fun resync() {
        if (!isPlaying) return
        if (kotlin.math.abs(shadow.currentPosition - active.currentPosition) > 750) {
            shadow.seekTo(active.currentPosition)
        }
    }

    private fun applyVolumes() {
        original.volume = if (useRemaster) 0f else 1f
        remaster.volume = if (useRemaster) 1f else 0f
    }

    fun release() {
        original.release()
        remaster.release()
    }
}
